File: python_services/routes/roslyn_routes.py
"""
Roslyn routes for C# code analysis using Roslyn.

Provides endpoints for:
- C# code analysis with Roslyn
- Semantic search for methods, classes, and event handlers
- Call graph traversal and impact analysis
- Repository statistics
"""
import logging
import subprocess
import tempfile
from pathlib import Path
from typing import Optional, List

from fastapi import APIRouter, Query, HTTPException, Body

import sys
sys.path.insert(0, str(Path(__file__).parent.parent))

from config import COLLECTION_CODE_CALLGRAPH, COLLECTION_CODE_METHODS
from roslyn_mongodb_service import get_roslyn_mongodb_service

logger = logging.getLogger(__name__)

# Create router
router = APIRouter(prefix="/roslyn", tags=["Roslyn"])


# ============================================================================
# Roslyn Analysis Routes
# ============================================================================

@router.post("/analyze")
async def analyze_csharp_code(
    input_path: str = Body(..., description="Path to C# file or directory"),
    project: str = Body("Unknown", description="Project name for tagging"),
    store: bool = Body(True, description="Store results in MongoDB")
):
    """
    Analyze C# code using Roslyn and optionally store results in MongoDB.

    Extracts:
    - Classes with inheritance and members
    - Methods with signatures, complexity, and SQL operations
    - Call graph relationships
    - Event handlers (UI to code mapping)
    - Database operations
    """
    roslyn = get_roslyn_mongodb_service()
    await roslyn.initialize()

    # Run Roslyn analyzer
    result = await roslyn.analyze_code(input_path, project)

    if not result['success']:
        raise HTTPException(status_code=500, detail=result.get('error', 'Analysis failed'))

    if store:
        # Store in MongoDB with embeddings
        stats = await roslyn.store_analysis(result)
        return {
            "success": True,
            "project": project,
            "input_path": input_path,
            "storage_stats": stats,
            "data_summary": {
                "classes": len(result['data'].get('Classes', [])),
                "methods": len(result['data'].get('Methods', [])),
                "call_graph": len(result['data'].get('CallGraph', [])),
                "event_handlers": len(result['data'].get('EventHandlers', [])),
                "db_operations": len(result['data'].get('DatabaseOperations', []))
            }
        }
    else:
        return {
            "success": True,
            "project": project,
            "input_path": input_path,
            "data": result['data']
        }


@router.get("/search/methods")
async def search_roslyn_methods(
    query: str = Query(..., description="Natural language query"),
    project: Optional[str] = Query(None, description="Filter by project"),
    limit: int = Query(10, ge=1, le=100),
    sql_only: bool = Query(False, description="Only return methods with SQL operations")
):
    """
    Search for C# methods using semantic similarity.

    Examples:
    - "methods that save bale data"
    - "button click handlers"
    - "methods that call stored procedures"
    """
    roslyn = get_roslyn_mongodb_service()
    await roslyn.initialize()

    results = await roslyn.search_methods(query, project, limit, sql_only)
    return {
        "success": True,
        "query": query,
        "count": len(results),
        "results": results
    }


@router.get("/search/classes")
async def search_roslyn_classes(
    query: str = Query(..., description="Natural language query"),
    project: Optional[str] = Query(None, description="Filter by project"),
    limit: int = Query(10, ge=1, le=100)
):
    """
    Search for C# classes using semantic similarity.

    Examples:
    - "view model for bale entry"
    - "classes that handle shipping orders"
    """
    roslyn = get_roslyn_mongodb_service()
    await roslyn.initialize()

    results = await roslyn.search_classes(query, project, limit)
    return {
        "success": True,
        "query": query,
        "count": len(results),
        "results": results
    }


@router.get("/search/event-handlers")
async def search_roslyn_event_handlers(
    query: str = Query(..., description="Natural language query"),
    project: Optional[str] = Query(None, description="Filter by project"),
    limit: int = Query(10, ge=1, le=100)
):
    """
    Search for event handlers (UI to code mapping).

    Examples:
    - "what happens when Save button is clicked"
    - "button click handlers for shipping"
    """
    roslyn = get_roslyn_mongodb_service()
    await roslyn.initialize()

    results = await roslyn.search_event_handlers(query, project, limit)
    return {
        "success": True,
        "query": query,
        "count": len(results),
        "results": results
    }


@router.get("/call-chain")
async def get_roslyn_call_chain(
    method_name: str = Query(..., description="Method name"),
    class_name: str = Query(..., description="Class name"),
    project: Optional[str] = Query(None, description="Filter by project"),
    direction: str = Query("both", description="callers, callees, or both"),
    max_depth: int = Query(3, ge=1, le=10, description="Max traversal depth")
):
    """
    Get the call chain for a method.

    Returns who calls this method and/or what it calls.
    Useful for understanding code flow and impact analysis.
    """
    roslyn = get_roslyn_mongodb_service()
    await roslyn.initialize()

    result = await roslyn.get_call_chain(method_name, class_name, project, direction, max_depth)
    return {
        "success": True,
        **result
    }


@router.get("/stats")
async def get_roslyn_stats():
    """Get statistics for all Roslyn collections in MongoDB."""
    roslyn = get_roslyn_mongodb_service()
    await roslyn.initialize()

    stats = await roslyn.get_stats()
    return {
        "success": True,
        **stats
    }


@router.post("/analyze-targeted")
async def analyze_targeted_files(
    methods: List[str] = Body(..., description="List of method names to find references for"),
    repo_path: str = Body(..., description="Path to repository root"),
    project: str = Body(..., description="Project name")
):
    """
    Targeted analysis: Find files containing method references and analyze only those.

    This is much faster than full repository analysis when you only need to update
    call graph data for specific methods.

    Steps:
    1. Search repository for files containing method name references
    2. Create a file list of matching C# files
    3. Run Roslyn analyzer on just those files using --filelist mode
    4. Store updated call graph data
    """
    import os

    roslyn = get_roslyn_mongodb_service()
    await roslyn.initialize()

    repo_path_obj = Path(repo_path)
    if not repo_path_obj.exists():
        raise HTTPException(status_code=400, detail=f"Repository path does not exist: {repo_path}")

    # Find files containing references to the methods
    files_to_analyze = set()

    for method in methods:
        # Extract just the method name if it's in ClassName.MethodName format
        method_name = method.split('.')[-1] if '.' in method else method

        try:
            # Use grep to find files containing the method name
            result = subprocess.run(
                ['grep', '-rl', '--include=*.cs', method_name, str(repo_path_obj)],
                capture_output=True,
                text=True,
                timeout=30
            )

            if result.stdout:
                for file_path in result.stdout.strip().split('\n'):
                    if file_path and file_path.endswith('.cs'):
                        # Exclude obj and bin directories
                        if '\\obj\\' not in file_path and '/obj/' not in file_path:
                            if '\\bin\\' not in file_path and '/bin/' not in file_path:
                                files_to_analyze.add(file_path)
        except subprocess.TimeoutExpired:
            print(f"Grep timeout for method: {method_name}")
        except Exception as e:
            print(f"Error searching for {method_name}: {e}")

    if not files_to_analyze:
        return {
            "success": True,
            "message": "No files found containing method references",
            "methods_searched": methods,
            "files_analyzed": 0
        }

    # Create a temporary file list for Roslyn analyzer
    with tempfile.NamedTemporaryFile(mode='w', suffix='.txt', delete=False) as f:
        for file_path in files_to_analyze:
            f.write(file_path + '\n')
        file_list_path = f.name

    try:
        # Run Roslyn analyzer with --filelist mode
        analysis_result = await roslyn.analyze_code_with_filelist(file_list_path, project)

        if analysis_result.get('success'):
            # Store results in MongoDB
            stats = await roslyn.store_analysis(analysis_result)

            return {
                "success": True,
                "project": project,
                "methods_searched": methods,
                "files_found": len(files_to_analyze),
                "files_analyzed": list(files_to_analyze)[:20],  # Limit for response size
                "files_truncated": len(files_to_analyze) > 20,
                "storage_stats": stats,
                "data_summary": {
                    "classes": len(analysis_result['data'].get('Classes', [])),
                    "methods": len(analysis_result['data'].get('Methods', [])),
                    "call_graph": len(analysis_result['data'].get('CallGraph', []))
                }
            }
        else:
            return {
                "success": False,
                "error": analysis_result.get('error', 'Analysis failed'),
                "files_attempted": len(files_to_analyze)
            }
    finally:
        try:
            import os
            os.unlink(file_list_path)
        except:
            pass


@router.post("/analyze-full")
async def analyze_full_repository(
    repo_path: str = Body(..., description="Path to repository root"),
    project: str = Body(..., description="Project name"),
    generate_embeddings: bool = Body(True, description="Generate vector embeddings")
):
    """
    Full repository analysis: Analyze all C# files and build complete call graph.

    This is a comprehensive analysis that:
    1. Scans all .cs files in the repository
    2. Extracts classes, methods, and call relationships
    3. Generates embeddings for semantic search
    4. Stores everything in MongoDB

    Use this for initial setup or periodic full refresh.
    For incremental updates, use /roslyn/analyze-targeted instead.
    """
    roslyn = get_roslyn_mongodb_service()
    await roslyn.initialize()

    repo_path_obj = Path(repo_path)
    if not repo_path_obj.exists():
        raise HTTPException(status_code=400, detail=f"Repository path does not exist: {repo_path}")

    # Count files first
    cs_files = list(repo_path_obj.rglob('*.cs'))
    cs_files = [f for f in cs_files if '\\obj\\' not in str(f) and '/obj/' not in str(f)]
    cs_files = [f for f in cs_files if '\\bin\\' not in str(f) and '/bin/' not in str(f)]

    if not cs_files:
        raise HTTPException(status_code=400, detail="No C# files found in repository")

    print(f"Starting full analysis of {project}: {len(cs_files)} files")

    # Run full analysis
    analysis_result = await roslyn.analyze_code(str(repo_path_obj), project)

    if not analysis_result.get('success'):
        raise HTTPException(status_code=500, detail=analysis_result.get('error', 'Analysis failed'))

    # Store in MongoDB
    stats = await roslyn.store_analysis(analysis_result, generate_embeddings)

    return {
        "success": True,
        "project": project,
        "repo_path": repo_path,
        "total_files": len(cs_files),
        "storage_stats": stats,
        "data_summary": {
            "classes": len(analysis_result['data'].get('Classes', [])),
            "methods": len(analysis_result['data'].get('Methods', [])),
            "call_graph": len(analysis_result['data'].get('CallGraph', [])),
            "event_handlers": len(analysis_result['data'].get('EventHandlers', [])),
            "db_operations": len(analysis_result['data'].get('DatabaseOperations', []))
        }
    }


@router.get("/callgraph-status/{project}")
async def get_callgraph_status(project: str):
    """
    Check if call graph data exists for a project.

    Returns counts of call graph entries and when they were last updated.
    Useful for determining if Roslyn analysis needs to be run.
    """
    roslyn = get_roslyn_mongodb_service()
    await roslyn.initialize()

    # Get call graph count for this project
    callgraph_collection = roslyn.db[COLLECTION_CODE_CALLGRAPH]
    methods_collection = roslyn.db[COLLECTION_CODE_METHODS]

    callgraph_count = await callgraph_collection.count_documents({"project": project})
    methods_count = await methods_collection.count_documents({"project": project})

    # Get most recent entry timestamp
    latest_entry = await callgraph_collection.find_one(
        {"project": project},
        sort=[("timestamp", -1)]
    )

    last_updated = latest_entry.get("timestamp") if latest_entry else None

    return {
        "success": True,
        "project": project,
        "has_callgraph": callgraph_count > 0,
        "callgraph_entries": callgraph_count,
        "methods_indexed": methods_count,
        "last_updated": last_updated,
        "needs_analysis": callgraph_count == 0
    }

File: src/routes/roslynRoutes.js
/**
 * Roslyn Routes Proxy
 *
 * Thin HTTP layer that proxies all Roslyn C# code analysis operations to the Python FastAPI service.
 *
 * The Python service handles:
 * - C# code analysis using Roslyn compiler APIs
 * - Storage in MongoDB with vector embeddings
 * - Semantic search for classes, methods, and event handlers
 * - Call chain analysis (who calls what)
 * - Database operation tracking
 *
 * Node.js only handles:
 * - HTTP routing
 * - Authentication middleware
 * - Request/response formatting
 */

import express from 'express';

const PYTHON_SERVICE_URL = process.env.PYTHON_SERVICE_URL || 'http://localhost:8001';

/**
 * Factory function to create Roslyn proxy routes
 *
 * @param {Object} deps - Dependencies
 * @param {Object} deps.monitoring - Production monitoring instance
 * @param {Function} deps.requireAuth - Authentication middleware
 * @returns {express.Router} Express router with all Roslyn proxy routes
 */
export default function createRoslynRoutes(deps) {
  const router = express.Router();

  const {
    monitoring,
    requireAuth
  } = deps;

  // Health check for Python service connection
  let pythonServiceAvailable = false;

  async function checkPythonService() {
    try {
      const response = await fetch(`${PYTHON_SERVICE_URL}/status`);
      pythonServiceAvailable = response.ok;
    } catch (error) {
      pythonServiceAvailable = false;
      monitoring.warn('Python Roslyn Service not available:', error.message);
    }
  }

  // Check on startup
  checkPythonService();

  // ============================================================================
  // Roslyn Analysis Routes (Proxied to Python)
  // ============================================================================

  /**
   * POST /api/roslyn/analyze
   * Analyze C# code using Roslyn and store results in MongoDB
   */
  router.post('/analyze', requireAuth, async (req, res) => {
    try {
      if (!pythonServiceAvailable) {
        await checkPythonService();
      }

      const { input_path, project, store = true } = req.body;

      if (!input_path) {
        return res.status(400).json({
          success: false,
          error: 'input_path is required'
        });
      }

      monitoring.info(`Analyzing C# code: ${input_path} (project: ${project || 'Unknown'})`);

      const response = await fetch(`${PYTHON_SERVICE_URL}/roslyn/analyze`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ input_path, project, store })
      });

      const result = await response.json();

      if (!response.ok) {
        return res.status(response.status).json(result);
      }

      res.json(result);

    } catch (error) {
      monitoring.error('Failed to analyze C# code:', error.message);
      res.status(500).json({
        success: false,
        error: 'Failed to analyze C# code',
        details: error.message
      });
    }
  });

  /**
   * POST /api/roslyn/call-chain
   * Get the call chain for a method (who calls it, what it calls)
   */
  router.post('/call-chain', async (req, res) => {
    try {
      const { className, methodName, project, direction = 'callers', max_depth = 3 } = req.body;

      if (!className || !methodName) {
        return res.status(400).json({
          success: false,
          error: 'className and methodName are required'
        });
      }

      monitoring.info(`Getting call chain for ${className}.${methodName}`);

      const response = await fetch(`${PYTHON_SERVICE_URL}/roslyn/call-chain`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ className, methodName, project, direction, max_depth })
      });

      const result = await response.json();

      if (!response.ok) {
        return res.status(response.status).json(result);
      }

      res.json(result);

    } catch (error) {
      monitoring.error('Failed to get call chain:', error.message);
      res.status(500).json({
        success: false,
        error: 'Failed to get call chain',
        details: error.message
      });
    }
  });

  /**
   * GET /api/roslyn/call-chain
   * Get the call chain for a method (query parameter version)
   */
  router.get('/call-chain', async (req, res) => {
    try {
      const { method_name, class_name, project, direction = 'both', max_depth = 3 } = req.query;

      if (!class_name || !method_name) {
        return res.status(400).json({
          success: false,
          error: 'class_name and method_name query parameters are required'
        });
      }

      const params = new URLSearchParams({
        method_name,
        class_name,
        direction,
        max_depth: max_depth.toString()
      });

      if (project) {
        params.append('project', project);
      }

      const response = await fetch(`${PYTHON_SERVICE_URL}/roslyn/call-chain?${params}`);
      const result = await response.json();

      if (!response.ok) {
        return res.status(response.status).json(result);
      }

      res.json(result);

    } catch (error) {
      monitoring.error('Failed to get call chain:', error.message);
      res.status(500).json({
        success: false,
        error: error.message
      });
    }
  });

  /**
   * POST /api/roslyn/search-methods
   * Search for C# methods using semantic similarity
   */
  router.post('/search-methods', async (req, res) => {
    try {
      const { query, project, limit = 10, includeSqlOnly = false } = req.body;

      if (!query) {
        return res.status(400).json({
          success: false,
          error: 'query is required'
        });
      }

      monitoring.info(`Searching methods: "${query}"`);

      const response = await fetch(`${PYTHON_SERVICE_URL}/roslyn/search-methods`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ query, project, limit, includeSqlOnly })
      });

      const result = await response.json();

      if (!response.ok) {
        return res.status(response.status).json(result);
      }

      res.json(result);

    } catch (error) {
      monitoring.error('Failed to search methods:', error.message);
      res.status(500).json({
        success: false,
        error: error.message
      });
    }
  });

  /**
   * GET /api/roslyn/search/methods
   * Search for C# methods using semantic similarity (GET version)
   */
  router.get('/search/methods', async (req, res) => {
    try {
      const { query, project, limit = 10, sql_only = false } = req.query;

      if (!query) {
        return res.status(400).json({
          success: false,
          error: 'query parameter is required'
        });
      }

      const params = new URLSearchParams({
        query,
        limit: limit.toString(),
        sql_only: sql_only.toString()
      });

      if (project) {
        params.append('project', project);
      }

      const response = await fetch(`${PYTHON_SERVICE_URL}/roslyn/search/methods?${params}`);
      const result = await response.json();

      if (!response.ok) {
        return res.status(response.status).json(result);
      }

      res.json(result);

    } catch (error) {
      monitoring.error('Failed to search methods:', error.message);
      res.status(500).json({
        success: false,
        error: error.message
      });
    }
  });

  /**
   * GET /api/roslyn/search/classes
   * Search for C# classes using semantic similarity
   */
  router.get('/search/classes', async (req, res) => {
    try {
      const { query, project, limit = 10 } = req.query;

      if (!query) {
        return res.status(400).json({
          success: false,
          error: 'query parameter is required'
        });
      }

      const params = new URLSearchParams({
        query,
        limit: limit.toString()
      });

      if (project) {
        params.append('project', project);
      }

      const response = await fetch(`${PYTHON_SERVICE_URL}/roslyn/search/classes?${params}`);
      const result = await response.json();

      if (!response.ok) {
        return res.status(response.status).json(result);
      }

      res.json(result);

    } catch (error) {
      monitoring.error('Failed to search classes:', error.message);
      res.status(500).json({
        success: false,
        error: error.message
      });
    }
  });

  /**
   * GET /api/roslyn/search/event-handlers
   * Search for event handlers (UI to code mapping)
   */
  router.get('/search/event-handlers', async (req, res) => {
    try {
      const { query, project, limit = 10 } = req.query;

      if (!query) {
        return res.status(400).json({
          success: false,
          error: 'query parameter is required'
        });
      }

      const params = new URLSearchParams({
        query,
        limit: limit.toString()
      });

      if (project) {
        params.append('project', project);
      }

      const response = await fetch(`${PYTHON_SERVICE_URL}/roslyn/search/event-handlers?${params}`);
      const result = await response.json();

      if (!response.ok) {
        return res.status(response.status).json(result);
      }

      res.json(result);

    } catch (error) {
      monitoring.error('Failed to search event handlers:', error.message);
      res.status(500).json({
        success: false,
        error: error.message
      });
    }
  });

  /**
   * GET /api/roslyn/stats
   * Get statistics for all Roslyn collections
   */
  router.get('/stats', async (req, res) => {
    try {
      const response = await fetch(`${PYTHON_SERVICE_URL}/roslyn/stats`);
      const result = await response.json();

      if (!response.ok) {
        return res.status(response.status).json(result);
      }

      res.json(result);

    } catch (error) {
      monitoring.error('Failed to get Roslyn stats:', error.message);
      res.status(500).json({
        success: false,
        error: error.message
      });
    }
  });

  // ============================================================================
  // Targeted and Full Analysis Routes
  // ============================================================================

  /**
   * POST /api/roslyn/analyze-targeted
   * Analyze specific files containing references to modified methods.
   * Finds files via string search, then runs Roslyn analysis only on those files.
   */
  router.post('/analyze-targeted', requireAuth, async (req, res) => {
    try {
      const { methods, repoPath, project } = req.body;

      if (!methods || !Array.isArray(methods) || methods.length === 0) {
        return res.status(400).json({
          success: false,
          error: 'methods array is required'
        });
      }

      if (!repoPath || !project) {
        return res.status(400).json({
          success: false,
          error: 'repoPath and project are required'
        });
      }

      monitoring.info(`Targeted analysis: ${methods.length} methods in ${project}`);

      // Proxy to Python service which will:
      // 1. Search for files containing method references
      // 2. Run Roslyn analyzer on just those files
      // 3. Store call graph results
      const response = await fetch(`${PYTHON_SERVICE_URL}/roslyn/analyze-targeted`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ methods, repo_path: repoPath, project })
      });

      const result = await response.json();

      if (!response.ok) {
        return res.status(response.status).json(result);
      }

      res.json(result);

    } catch (error) {
      monitoring.error('Failed targeted analysis:', error.message);
      res.status(500).json({
        success: false,
        error: 'Failed to run targeted analysis',
        details: error.message
      });
    }
  });

  /**
   * POST /api/roslyn/analyze-full
   * Full repository analysis using Roslyn.
   * Analyzes all C# files and stores complete call graph in MongoDB.
   */
  router.post('/analyze-full', requireAuth, async (req, res) => {
    try {
      const { repoPath, project, generateEmbeddings = true } = req.body;

      if (!repoPath || !project) {
        return res.status(400).json({
          success: false,
          error: 'repoPath and project are required'
        });
      }

      monitoring.info(`Full repository analysis: ${project} at ${repoPath}`);

      // Proxy to Python service for full analysis
      const response = await fetch(`${PYTHON_SERVICE_URL}/roslyn/analyze-full`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          repo_path: repoPath,
          project,
          generate_embeddings: generateEmbeddings
        })
      });

      const result = await response.json();

      if (!response.ok) {
        return res.status(response.status).json(result);
      }

      res.json(result);

    } catch (error) {
      monitoring.error('Failed full analysis:', error.message);
      res.status(500).json({
        success: false,
        error: 'Failed to run full repository analysis',
        details: error.message
      });
    }
  });

  /**
   * GET /api/roslyn/callgraph-status/:project
   * Check if call graph data exists for a project
   */
  router.get('/callgraph-status/:project', async (req, res) => {
    try {
      const { project } = req.params;

      const response = await fetch(`${PYTHON_SERVICE_URL}/roslyn/callgraph-status/${encodeURIComponent(project)}`);
      const result = await response.json();

      if (!response.ok) {
        return res.status(response.status).json(result);
      }

      res.json(result);

    } catch (error) {
      monitoring.error('Failed to get call graph status:', error.message);
      res.status(500).json({
        success: false,
        error: error.message
      });
    }
  });

  // ============================================================================
  // Health Check
  // ============================================================================

  /**
   * GET /api/roslyn/health
   * Check Python Roslyn Service availability
   */
  router.get('/health', async (req, res) => {
    try {
      await checkPythonService();
      res.json({
        success: true,
        pythonServiceAvailable,
        message: pythonServiceAvailable
          ? 'Roslyn service connected to Python backend'
          : 'Python backend unavailable - Roslyn features may not work'
      });
    } catch (error) {
      res.status(503).json({
        success: false,
        pythonServiceAvailable: false,
        error: error.message
      });
    }
  });

  return router;
}
